active/0000-stateful-helpers.md

```diff
@@ -0,0 +1,322 @@
+- Start Date: 2015-05-17
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Stateful helpers are a class-based way to define helpers. Stateful helpers:
+
+  * Have a single return value
+  * Can store and read state
+  * Have lifecycle hooks analagous to components where appropriate. For
+    example, a stateful helper may call `recompute` at any time to generate a new
+    value.
+  * Are a superset of traditional function-based helpers. They can do more,
+    but in many cases a traditional helper is appropriate.
+
+They fill a gap in Ember's current template APIs:
+
+|                  | has positional params | has layout | can yield template | has lifecycle, instance | can control rerender |
+|------------------|-----------------------|------------|--------------------|-------------------------|----------------------|
+| components       | Yes                   | Yes        | Yes                | Yes                     | Yes                  |
+| stateful helpers | Yes                   | No         | Yes                | Yes                     | Yes                  |
+| function helpers | Yes                   | No         | Yes                | No                      | No                   |
+
+Example usage:
+
+```js
+// app/helpers/full-name.js
+import Ember from "ember";
```

Comment:
As another side note, I think we should be exploring using decorators to make APIs like this nicer:

export default Ember.Helper.extend({
  @recompute({ on: 'change' })
  nameBuilder: Ember.inject.service(),
  compute(params) {
    const builder = this.get('nameBuilder');
    return builder.fullName(params[0], params[1]);
  }
});